src/Validators/SwaggerSpecValidator.php

```diff
@@ -0,0 +1,442 @@
+                $validTypes = self::SCHEMA_TYPES;
+                break;
+            case 'formData':
+                $this->validateFormDataConsumes($operation, $operationId);
+
+                $requiredFields = ['type'];
+                $validTypes = array_merge(self::PRIMITIVE_TYPES, ['file']);
+                break;
+            default:
+                $requiredFields = ['type'];
+                $validTypes = self::PRIMITIVE_TYPES;
+        }
+
+        $this->validateFieldsPresent($requiredFields, $paramId);
+
+        $schema = Arr::get($param, 'schema', $param);
+        $this->validateType($schema, $validTypes, $paramId);
+    }
+
+    protected function validateHeader(array $header, string $headerId): void
+    {
+        $this->validateFieldsPresent(self::REQUIRED_FIELDS['header'], $headerId);
+
+        $this->validateType($header, self::PRIMITIVE_TYPES, $headerId);
+        $this->validateFieldValue("{$headerId}.collectionFormat", self::ALLOWED_VALUES['header_collection_format']);
+
+        if (!empty($header['items'])) {
+            $this->validateItems($header['items'], $headerId);
+        }
+    }
+
+    protected function validateItems(array $items, string $itemsId): void
+    {
+        $this->validateFieldsPresent(self::REQUIRED_FIELDS['item'], $itemsId);
+
+        $this->validateType($items, self::PRIMITIVE_TYPES, $itemsId);
+        $this->validateFieldValue("{$itemsId}.collectionFormat", self::ALLOWED_VALUES['items_collection_format']);
+    }
+
+    protected function getMissingFields(array $requiredFields, ?string $fieldId = null, ?array $doc = null): array
+    {
+        return array_diff(
+            $requiredFields,
+            array_keys(Arr::get($doc ?? $this->doc, $fieldId))
+        );
+    }
+
+    protected function validateFieldsPresent(array $requiredFields, ?string $fieldId = null): void
+    {
+        $missingDocFields = $this->getMissingFields($requiredFields, $fieldId);
+
+        if (!empty($missingDocFields)) {
+            throw new MissingFieldException($missingDocFields, $fieldId);
+        }
+    }
+
+    protected function validateFieldValue(string $fieldId, array $allowedValues): void
+    {
+        $inputValue = Arr::wrap(Arr::get($this->doc, $fieldId, []));
+        $approvedValues = array_intersect($inputValue, $allowedValues);
+        $invalidValues = array_diff($inputValue, $approvedValues);
+
+        if (!empty($invalidValues)) {
+            throw new InvalidFieldValueException($fieldId, $allowedValues);
+        }
+    }
+
+    protected function validateRefs(): void
+    {
+        array_walk_recursive($this->doc, function ($item, $key) {
+            if (
+                ($key === '$ref')
+                && preg_match(self::REF_REGEXP, $item, $matches)
+            ) {
+                $refFilename = Arr::get($matches, 1);
+                $refParentKey = Arr::get($matches, 2);
+                $refKey = Arr::get($matches, 3);
```

Comment:
```suggestion
                list($other, $refFilename, $refParentKey, $refKey) = $matches;
```

Reply:
That's not possible because sometimes there is no 2nd or 3rd values.